test_ls372_set_heater_range is flaky

The LS372 integration test, `test_ls372_set_heater_range`, can be flaky sometimes. The communication times out, which seems like it might be an issue on the simulator end.

Now that the TCP relay exists in the device emulator it would simplify the testing here if use of the simulator was replaced with the device emulator.

## Logs
### pytest
```
 ============================= test session starts ==============================
platform linux -- Python 3.8.12, pytest-7.0.1, pluggy-1.0.0
rootdir: /home/runner/work/socs/socs/tests, configfile: pytest.ini
plugins: docker-compose-3.2.1, dependency-0.5.1, cov-3.0.0, order-1.0.1
collected 98 items / 67 deselected / 31 selected

integration/test_ls240_agent_integration.py ....                         [ 12%]
integration/test_ls372_agent_integration.py .....F..............         [ 77%]
integration/test_ls425_agent_integration.py .......                      [100%]

=================================== FAILURES ===================================
_________________________ test_ls372_set_heater_range __________________________

wait_for_crossbar = None, run_agent = None
client = <ocs.ocs_client.OCSClient object at 0x7f724c820520>

    @pytest.mark.integtest
    def test_ls372_set_heater_range(wait_for_crossbar, run_agent, client):
        client.init_lakeshore()
        resp = client.set_heater_range(range=1e-3, heater='sample', wait=0)
        assert resp.status == ocs.OK
>       assert resp.session['op_code'] == OpCode.SUCCEEDED.value
E       assert 6 == 5
E        +  where 5 = <OpCode.SUCCEEDED: 5>.value
E        +    where <OpCode.SUCCEEDED: 5> = OpCode.SUCCEEDED

integration/test_ls372_agent_integration.py:72: AssertionError
---------------------------- Captured stdout setup -----------------------------
```

### Simulator
```
2022-02-15 20:47:06,698 [INFO]:  Received command: INSET? 16
2022-02-15 20:47:06,698 [INFO]:  Sent response: 1,10,3,0,1
2022-02-15 20:47:06,698 [INFO]:  Received command: INNAME? 16
2022-02-15 20:47:06,698 [INFO]:  Sent response: Channel 16
2022-02-15 20:47:06,698 [INFO]:  Received command: INTYPE? 16
2022-02-15 20:47:06,698 [INFO]:  Sent response: 0,1,0,1,0,1
2022-02-15 20:47:06,699 [INFO]:  Received command: TLIMIT? 16
2022-02-15 20:47:06,699 [INFO]:  Sent response: 0
2022-02-15 20:47:06,699 [INFO]:  Received command: OUTMODE? 0
2022-02-15 20:47:06,699 [INFO]:  Sent response: 0,1,0,0,0,5
2022-02-15 20:47:06,699 [INFO]:  Received command: HTRSET? 0
2022-02-15 20:47:06,699 [INFO]:  Sent response: 1,0,0,1
2022-02-15 20:47:06,699 [INFO]:  Received command: OUTMODE? 2
2022-02-15 20:47:06,699 [INFO]:  Sent response: 0,1,0,0,0,5
2022-02-15 20:47:06,699 [INFO]:  Received command: HTRSET? 2
2022-02-15 20:47:06,699 [INFO]:  Sent response: 1,0,0,1
2022-02-15 20:47:07,163 [INFO]:  Received command: RANGE? 0
2022-02-15 20:47:07,163 [INFO]:  Sent response: 0
2022-02-15 20:47:07,164 [INFO]:  Received command: RANGE 0 4
RANGE? 0
2022-02-15 20:47:07,164 [ERROR]:  Command error: set_heater_range() missing 1 required positional argument: 'heater_range'
2022-02-15 20:47:27,350 [INFO]:  Connection closed by client
2022-02-15 20:47:27,350 [INFO]:  waiting for a connection....
2022-02-15 20:47:28,883 [INFO]:  Made connection with ('192.168.192.1', 60978)
2022-02-15 20:47:28,883 [INFO]:  Received command: *IDN?
2022-02-15 20:47:28,883 [INFO]:  Sent response: LSCI,MODEL372,LSASIM,0.0
```

### Agent
```
2022-02-15T20-59-15.693083 set_heater_range:2 Status is now "starting".
2022-02-15T20-59-15.693965 set_heater_range:2 Status is now "running".
2022-02-15T20-59-25.702503 Warning: Caught timeout waiting for response to 'RANGE? 0', trying again before giving up
2022-02-15T20-59-35.712146 Warning: Caught timeout waiting for response to 'RANGE? 0', trying again before giving up
2022-02-15T20-59-35.715265 set_heater_range:2 CRASH: [Failure instance: Traceback: <class 'RuntimeError'>: Query response to Lakeshore timed out after two attempts. Check connection.
/usr/lib/python3.10/threading.py:1009:_bootstrap_inner
/usr/lib/python3.10/threading.py:946:run
/home/koopman/.local/lib/python3.10/site-packages/twisted/_threads/_threadworker.py:47:work
/home/koopman/.local/lib/python3.10/site-packages/twisted/_threads/_team.py:181:doWork
--- <exception caught here> ---
/home/koopman/.local/lib/python3.10/site-packages/twisted/python/threadpool.py:238:inContext
/home/koopman/.local/lib/python3.10/site-packages/twisted/python/threadpool.py:254:<lambda>
/home/koopman/.local/lib/python3.10/site-packages/twisted/python/context.py:118:callWithContext
/home/koopman/.local/lib/python3.10/site-packages/twisted/python/context.py:83:callWithContext
/home/koopman/git/socs/agents/lakeshore372/LS372_agent.py:444:set_heater_range
/home/koopman/git/socs/socs/Lakeshore/Lakeshore372.py:1735:set_heater_range
/home/koopman/git/socs/socs/Lakeshore/Lakeshore372.py:1743:get_heater_range
/home/koopman/git/socs/socs/Lakeshore/Lakeshore372.py:267:msg
]
2022-02-15T20-59-35.717173 set_heater_range:2 Status is now "done".
2022-02-15T20-59-35.773073 Received SIGINT, shutting down.
2022-02-15T20-59-35.774638 session left: CloseDetails(reason=<wamp.close.normal>, message='None')
2022-02-15T20-59-35.774821 Stopping session init_lakeshore
```

--- a/.github/workflows/deploy.yml
+++ b/.github/workflows/deploy.yml
@@ -47,10 +47,6 @@ jobs:
         docker-compose build socs
 
     # Integration Tests
-    - name: Build images for integration tests
-      run: |
-        docker-compose build ocs-lakeshore372-simulator
-
     - name: Run integration tests
       working-directory: ./tests
       run: |

--- a/.github/workflows/develop.yml
+++ b/.github/workflows/develop.yml
@@ -52,10 +52,6 @@ jobs:
         docker-compose build socs
 
     # Integration Tests
-    - name: Build images for integration tests
-      run: |
-        docker-compose build ocs-lakeshore372-simulator
-
     - name: Run integration tests
       working-directory: ./tests
       run: |

--- a/.github/workflows/pytest.yml
+++ b/.github/workflows/pytest.yml
@@ -49,10 +49,6 @@ jobs:
         docker-compose build socs
 
     # Integration Tests
-    - name: Build images for integration tests
-      run: |
-        docker-compose build ocs-lakeshore372-simulator
-
     - name: Run integration tests
       working-directory: ./tests
       run: |

--- a/tests/default.yaml
+++ b/tests/default.yaml
@@ -16,7 +16,7 @@ hosts:
                      ['--ip-address', '127.0.0.1'],
                      ['--dwell-time-delay', 0],
                      ['--sample-heater', False],
-                     ['--mode', 'init']]},
+                     ['--mode', 'idle']]},
       {'agent-class': 'Lakeshore425Agent',
       'instance-id': 'LS425',
       'arguments': [

--- a/tests/docker-compose.yml
+++ b/tests/docker-compose.yml
@@ -8,8 +8,3 @@ services:
       - ./config.json:/app/crossbar/config.json
     environment:
          - PYTHONUNBUFFERED=1
-
-  ocs-LS372:
-    image: ocs-lakeshore372-simulator
-    ports:
-      - "7777:7777"

--- a/tests/integration/test_ls372_agent_integration.py
+++ b/tests/integration/test_ls372_agent_integration.py
@@ -7,6 +7,8 @@ from ocs.testing import create_agent_runner_fixture, create_client_fixture
 
 from integration.util import create_crossbar_fixture
 
+from socs.testing.device_emulator import create_device_emulator
+
 pytest_plugins = ("docker_compose")
 
 # Set the OCS_CONFIG_DIR so we read the local default.yaml file always
@@ -19,6 +21,40 @@ run_agent = create_agent_runner_fixture(
 client = create_client_fixture('LSASIM')
 wait_for_crossbar = create_crossbar_fixture()
 
+def build_init_responses():
+    values = {'*IDN?': 'LSCI,MODEL372,LSASIM,1.3',
+              'SCAN?': '01,1',
+              'INSET? A': '0,010,003,00,1',
+              'INNAME? A': 'Input A',
+              'INTYPE? A': '1,04,0,15,0,2',
+              'TLIMIT? A': '+0000',
+              'OUTMODE? 0': '2,6,1,0,0,001',
+              'HTRSET? 0': '+120.000,8,+0000.00,1',
+              'OUTMODE? 2': '4,16,1,0,0,001',
+              'HTRSET? 2': '+120.000,8,+0000.00,1'}
+
+    for i in range(1, 17):
+        values[f'INSET? {i}'] = '1,007,003,21,1'
+        values[f'INNAME? {i}'] = f'Channel {i:02}'
+        values[f'INTYPE? {i}'] = '0,07,1,10,0,1'
+        values[f'TLIMIT? {i}'] = '+0000'
+
+    # Heaters
+    values.update({'RANGE? 0': '0',
+                   'RANGE? 2': '1',
+                   'STILL?': '+10.60',
+                   'HTR?': '+00.0005E+00'})
+
+    # Senor readings
+    values.update({'KRDG? 1': '+293.873E+00',
+                   'SRDG? 1': '+108.278E+00',
+                   'KRDG? A': '+00.0000E-03',
+                   'SRDG? A': '+000.000E+09'})
+
+    return values
+
+emulator = create_device_emulator(build_init_responses(), relay_type='tcp', port=7777)
+
 
 @pytest.mark.integtest
 def test_testing(wait_for_crossbar):
@@ -27,7 +63,7 @@ def test_testing(wait_for_crossbar):
 
 
 @pytest.mark.integtest
-def test_ls372_init_lakeshore(wait_for_crossbar, run_agent, client):
+def test_ls372_init_lakeshore(wait_for_crossbar, emulator, run_agent, client):
     resp = client.init_lakeshore()
     # print(resp)
     assert resp.status == ocs.OK
@@ -36,7 +72,7 @@ def test_ls372_init_lakeshore(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_enable_control_chan(wait_for_crossbar, run_agent, client):
+def test_ls372_enable_control_chan(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.enable_control_chan()
     assert resp.status == ocs.OK
@@ -44,7 +80,7 @@ def test_ls372_enable_control_chan(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_disable_control_chan(wait_for_crossbar, run_agent, client):
+def test_ls372_disable_control_chan(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.disable_control_chan()
     assert resp.status == ocs.OK
@@ -52,7 +88,7 @@ def test_ls372_disable_control_chan(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_start_acq(wait_for_crossbar, run_agent, client):
+def test_ls372_start_acq(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.acq.start(sample_heater=False, run_once=True)
     assert resp.status == ocs.OK
@@ -65,7 +101,7 @@ def test_ls372_start_acq(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_heater_range(wait_for_crossbar, run_agent, client):
+def test_ls372_set_heater_range(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_heater_range(range=1e-3, heater='sample', wait=0)
     assert resp.status == ocs.OK
@@ -73,7 +109,7 @@ def test_ls372_set_heater_range(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_excitation_mode(wait_for_crossbar, run_agent, client):
+def test_ls372_set_excitation_mode(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_excitation_mode(channel=1, mode='current')
     assert resp.status == ocs.OK
@@ -81,7 +117,7 @@ def test_ls372_set_excitation_mode(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_excitation(wait_for_crossbar, run_agent, client):
+def test_ls372_set_excitation(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_excitation(channel=1, value=1e-9)
     assert resp.status == ocs.OK
@@ -89,17 +125,16 @@ def test_ls372_set_excitation(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_get_excitation(wait_for_crossbar, run_agent, client):
+def test_ls372_get_excitation(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
-    client.set_excitation(channel=1, value=1e-9)
     resp = client.get_excitation(channel=1)
     assert resp.status == ocs.OK
     assert resp.session['op_code'] == OpCode.SUCCEEDED.value
-    assert resp.session['data']['excitation'] == 1e-9
+    assert resp.session['data']['excitation'] == 2e-3
 
 
 @pytest.mark.integtest
-def test_ls372_set_resistance_range(wait_for_crossbar, run_agent, client):
+def test_ls372_set_resistance_range(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_resistance_range(channel=1, resistance_range=2)
     assert resp.status == ocs.OK
@@ -107,17 +142,17 @@ def test_ls372_set_resistance_range(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_get_resistance_range(wait_for_crossbar, run_agent, client):
+def test_ls372_get_resistance_range(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     client.set_resistance_range(channel=1, resistance_range=2)
     resp = client.get_resistance_range(channel=1)
     assert resp.status == ocs.OK
     assert resp.session['op_code'] == OpCode.SUCCEEDED.value
-    assert resp.session['data']['resistance_range'] == 2
+    assert resp.session['data']['resistance_range'] == 63.2
 
 
 @pytest.mark.integtest
-def test_ls372_set_dwell(wait_for_crossbar, run_agent, client):
+def test_ls372_set_dwell(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_dwell(channel=1, dwell=3)
     assert resp.status == ocs.OK
@@ -125,17 +160,16 @@ def test_ls372_set_dwell(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_get_dwell(wait_for_crossbar, run_agent, client):
+def test_ls372_get_dwell(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
-    client.set_dwell(channel=1, dwell=3)
     resp = client.get_dwell(channel=1)
     assert resp.status == ocs.OK
     assert resp.session['op_code'] == OpCode.SUCCEEDED.value
-    assert resp.session['data']['dwell_time'] == 3
+    assert resp.session['data']['dwell_time'] == 7
 
 
 @pytest.mark.integtest
-def test_ls372_set_pid(wait_for_crossbar, run_agent, client):
+def test_ls372_set_pid(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_pid(P=40, I=2, D=0)
     assert resp.status == ocs.OK
@@ -143,7 +177,7 @@ def test_ls372_set_pid(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_active_channel(wait_for_crossbar, run_agent, client):
+def test_ls372_set_active_channel(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_active_channel(channel=1)
     assert resp.status == ocs.OK
@@ -151,7 +185,7 @@ def test_ls372_set_active_channel(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_autoscan(wait_for_crossbar, run_agent, client):
+def test_ls372_set_autoscan(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_autoscan(autoscan=True)
     assert resp.status == ocs.OK
@@ -159,7 +193,7 @@ def test_ls372_set_autoscan(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_output_mode(wait_for_crossbar, run_agent, client):
+def test_ls372_set_output_mode(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_output_mode(heater='still', mode='Off')
     assert resp.status == ocs.OK
@@ -167,7 +201,7 @@ def test_ls372_set_output_mode(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_heater_output(wait_for_crossbar, run_agent, client):
+def test_ls372_set_heater_output(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_heater_output(heater='still', output=50)
     assert resp.status == ocs.OK
@@ -175,7 +209,7 @@ def test_ls372_set_heater_output(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_set_still_output(wait_for_crossbar, run_agent, client):
+def test_ls372_set_still_output(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.set_still_output(output=50)
     assert resp.status == ocs.OK
@@ -183,7 +217,7 @@ def test_ls372_set_still_output(wait_for_crossbar, run_agent, client):
 
 
 @pytest.mark.integtest
-def test_ls372_get_still_output(wait_for_crossbar, run_agent, client):
+def test_ls372_get_still_output(wait_for_crossbar, emulator, run_agent, client):
     client.init_lakeshore()
     resp = client.get_still_output()
     assert resp.status == ocs.OK

